lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+    def get_task_name_from_task(task_name)
+      task_name.rpartition('\\')[2]
+    end
+    
+  # Returns an array of scheduled task names.
+  # By default EVERYTHING is enumerated
+  # option hash
+  #    include_child_folders: recurses into child folders for tasks. Default true
+  #    include_compatibility: Only include tasks which have any of the specified compatibility levels. Default empty array (everything is permitted)
+  #
+  # DONE
+  def enum_task_names(folder_path = ROOT_FOLDER, options = nil)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless folder_path.is_a?(String)
+
+    options = {} if options.nil?
+    options[:include_child_folders] = true if options[:include_child_folders].nil?
+    options[:include_compatibility] = [] if options[:include_compatibility].nil?
+
+    array = []
+
+    task_folder = @task_service.GetFolder(folder_path)
+
+    task_folder.GetTasks(TASK_ENUM_HIDDEN).each do |task|
+      included = true
+
+      unless !included || options[:include_compatibility].empty?
+        included = included && options[:include_compatibility].include?(task.Definition.Settings.Compatibility)
+      end
+
+      array << task.Path if included
+    end
+    return array unless options[:include_child_folders]
+
+    task_folder.GetFolders(0).each do |child_folder|
+      array = array + enum_task_names(child_folder.Path, options)
+    end
+
+    array
+  end
+
+  # DONE
+  def activate(task)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless task.is_a?(String)
+
+    task_folder = @task_service.GetFolder(get_folder_path_from_task(task))
+
+    begin
+      @pITask = task_folder.GetTask(get_task_name_from_task(task))
+    rescue WIN32OLERuntimeError => e
+      @pITask = nil
+      # TODO win32ole errors are horrible.  Assume the task doesn't exist
+    end
+    @pITaskDefinition = nil
+    
+    @pITask
+  end
+
+  # DONE
+  def deactivate()
+    @pITask = nil
+    @pITaskDefinition = nil
+  end
+
+  # DONE
+  def definition()
+    if @pITaskDefinition.nil? && !@pITask.nil?
+      # Create a new editable Task Defintion based off of the currently activated task
+      @pITaskDefinition = @task_service.NewTask(0)
+      @pITaskDefinition.XmlText = @pITask.XML
+    end
+
+    @pITaskDefinition
+  end
+
+  # Delete the specified task name.
+  #
+  # DONE
+  def delete(task)
+    raise Error.new(_('No current task scheduler. Schedule.Service is NULL.')) if @task_service.nil?
+    raise TypeError unless task.is_a?(String)
+
+    task_folder = @task_service.GetFolder(get_folder_path_from_task(task))
+
+    result = -1
+    begin
+      result = task_folder.DeleteTask(get_task_name_from_task(task),0)
+    rescue WIN32OLERuntimeError => e
+      # TODO win32ole errors are horrible.  Assume the task doesn't exist so deletion is successful
+      return true
```

Comment:
For all the methods here, not sure that we should eat the `WIN32OLERuntimeError`.  In the prior implementation, nearly all COM errors end up calling `raise`. The `com.rb` file has a helper that automatically turns failed HRESULTs into exceptions.

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/com.rb#L140-L143

In the original v1, there are a few cases where well known HRESULTs are ignored during certain circumstances... whether they should be here or not remains to be seen.